### ISSUE TITLE
Semantics of `validate()` is unclear

### ISSUE THREAD

**Issue by mikand** [pyparsing/pyparsing#444]
Hi, 
thanks a lot for the great library! 

I am puzzled by the semantics of the `validate()` method: I tried with a very simple grammar (taken from https://github.com/pyparsing/pyparsing/blob/master/examples/simpleArith.py) and the `validate()` method fails:
```python
from pyparsing import *

integer = pyparsing_common.integer
variable = Word(alphas, exact=1)
operand = integer | variable

expop = Literal("^")
signop = oneOf("+ -")
multop = oneOf("* /")
plusop = oneOf("+ -")
factop = Literal("!")

expr = infixNotation(
    operand,
    [
        ("!", 1, opAssoc.LEFT),
        ("^", 2, opAssoc.RIGHT),
        (signop, 1, opAssoc.RIGHT),
        (multop, 2, opAssoc.LEFT),
        (plusop, 2, opAssoc.LEFT),
    ],
)

expr.validate()
```

Gives me:
```
pyparsing.exceptions.RecursiveGrammarException: RecursiveGrammarException: [+ | - term, + | - term, {FollowedBy>:({+ | - + | - term}) Group:({[+ | -] + | - term})}, Group:({[+ | -] + | - term}), {[+ | -] + | - term}, + | - term]
```

Is this an expected behavior? More in general, is it ok for a grammar to be `validate() == False`? The documentation is very unclear on this and I would like to understand if it is OK to create grammars that do not validate...

**Comment by ptmcg** [comment]
validate() is an old method, that I should probably deprecate, or at least update its docs. I wrote it when left-recursive (LR) grammars were not supported in pyparsing, so pyparsing grammars that were LR by accident ended up being just infinite loops with no real diagnostic on why.

validate() also can give false positives, of a grammar that might be problematic, but in practice is not.

In short, validate() is an outdated troubleshooting method, that is not really useful any more. Instead, use debugging methods like set_debug and railroad diagramming.

**Comment by ptmcg** [comment]
Released in 3.1.0a1
